PurchasesCoreSwift/Purchasing/StoreKitRequestFetcher.swift

```diff
@@ -0,0 +1,73 @@
+//
+//  StoreKitRequestFetcher.swift
+//  PurchasesCoreSwift
+//
+//  Created by Andrés Boedo on 6/29/21.
+//  Copyright © 2021 Purchases. All rights reserved.
+//
+
+import Foundation
+import StoreKit
+
+@objc(RCReceiptRefreshRequestFactory) public class ReceiptRefreshRequestFactory: NSObject {
+
+    @objc public func receiptRefreshRequest() -> SKReceiptRefreshRequest {
+        return SKReceiptRefreshRequest()
+    }
+}
+
+@objc(RCStoreKitRequestFetcher) public class StoreKitRequestFetcher: NSObject {
+    private let requestFactory: ReceiptRefreshRequestFactory
+    private var receiptRefreshRequest: SKRequest?
+    private var receiptRefreshCompletionHandlers: [() -> Void]
+    private let queue = DispatchQueue(label: "StoreKitRequestFetcher")
+
+    @objc public init(requestFactory: ReceiptRefreshRequestFactory = ReceiptRefreshRequestFactory()) {
+        self.requestFactory = requestFactory
+        receiptRefreshRequest = nil
+        receiptRefreshCompletionHandlers = []
+    }
+
+    @objc public func fetchReceiptData(_ completion: @escaping () -> Void) {
+        queue.async {
+            self.receiptRefreshCompletionHandlers.append(completion)
+
+            if self.receiptRefreshRequest == nil {
+                self.receiptRefreshRequest = self.requestFactory.receiptRefreshRequest()
+                self.receiptRefreshRequest?.delegate = self
+                self.receiptRefreshRequest?.start()
+            }
+        }
+    }
+}
+
+extension StoreKitRequestFetcher: SKRequestDelegate {
+
+    public func requestDidFinish(_ request: SKRequest) {
+        guard request is SKReceiptRefreshRequest else { return }
+
+        finishReceiptRequest(request)
+        request.cancel()
+    }
+
+    public func request(_ request: SKRequest, didFailWithError error: Error) {
+        guard request is SKReceiptRefreshRequest else { return }
+
+        Logger.appleError(String(format: Strings.offering.sk_request_failed, error.localizedDescription))
+        finishReceiptRequest(request)
+        request.cancel()
+    }
+}
+
+private extension StoreKitRequestFetcher {
+
+    func finishReceiptRequest(_ request: SKRequest?) {
+        queue.async {
+            self.receiptRefreshRequest = nil
+            for handler in self.receiptRefreshCompletionHandlers {
+                handler()
+            }
+            self.receiptRefreshCompletionHandlers = []
```

Comment:
another change in behavior here when comparing to obj-c: 
before, we were doing `@syncrhonized` to copy the completion handlers to a new variable, then resetting the  instance property, then calling the completion handlers. 

now, we're doing `queue.async` to copy the completion handlers to a new variable, then calling them, and only then resetting the variable and freeing up the queue. 

while I could just revert the order of those operations so that they look more similar, the important difference in behavior is that we're only freeing up the queue after the handlers are finished. 

I think this should be fine, but I'm going to test this out a bit more just in case before merging.